framework/include/utils/KDTree.h

```diff
@@ -0,0 +1,157 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef KDTREE_H
+#define KDTREE_H
+
+// Moose includes
+#include "MooseMesh.h"
+
+// Libmesh includes
+#include "libmesh/nanoflann.hpp"
+
+class KDTree
+{
+public:
+  KDTree(std::vector<Point> & master_points);
+
+  virtual ~KDTree() = default;
+
+  void buildTree();
+
+  void neighborSearch(Point & query_point,
+                      unsigned int patch_size,
+                      std::vector<unsigned int> & return_index);
+
+  /**
+   * PointListAdaptor is required to use libMesh Point coordinate type with
+   * nanoflann KDTree library. The member functions within the PointListAdaptor
+   * are used by nanoflann library.
+   */
+  template <unsigned int KDDim>
+  class PointListAdaptor
+  {
+  private:
+    const std::vector<Point> & _pts;
+
+  public:
+    PointListAdaptor(const std::vector<Point> & pts) : _pts(pts) {}
+
+    /**
+     * libMesh \p Point coordinate type
+     */
+    using coord_t = Real;
+    /**
+     * Must return the number of data points
+     */
+    inline size_t kdtree_get_point_count() const { return _pts.size(); }
+
+    /**
+     * Returns the distance between the vector "p1[0:size-1]"
+     * and the data point with index "idx_p2" stored in the class
+     */
+    inline coord_t kdtree_distance(const coord_t * p1, const size_t idx_p2, size_t size) const
+    {
+      mooseAssert(size == LIBMESH_DIM, "Size of point should equal libmesh dimension.");
+      mooseAssert(idx_p2 <= _pts.size(),
+                  "The point index should be less than"
+                  "total number of points used to build"
+                  "the KDTree.");
+
+      const Point & p2(_pts[idx_p2]);
+
+      switch (size)
```

Comment:
Yes, size has to be same as LIBMESH_DIM. I have removed the extra parameter as you suggested.